src/commands/close.rs

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) Radical HQ Limited
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+use indoc::formatdoc;
+
+use crate::{
+    error::{Error, Result},
+    github::{PullRequestState, PullRequestUpdate},
+    output::{output, write_commit_title},
+};
+
+pub async fn close(
+    git: &crate::git::Git,
+    gh: &mut crate::github::GitHub,
+    config: &crate::config::Config,
+) -> Result<()> {
+    let mut prepared_commits = git.get_prepared_commits(config)?;
+
+    let prepared_commit = match prepared_commits.last_mut() {
+        Some(c) => c,
+        None => {
+            output("👋", "Branch is empty - nothing to do. Good bye!")?;
+            return Ok(());
+        }
+    };
+
+    write_commit_title(prepared_commit)?;
+
+    let pull_request_number =
+        if let Some(number) = prepared_commit.pull_request_number {
+            output("#️⃣ ", &format!("Pull Request #{}", number))?;
+            number
+        } else {
+            return Err(Error::new(
+                "This commit does not refer to a Pull Request.",
+            ));
+        };
+
+    // Load Pull Request information
+    let pull_request = gh.get_pull_request(pull_request_number).await??;
+
+    if pull_request.state != PullRequestState::Open {
+        return Err(Error::new(formatdoc!(
+            "This Pull Request is already closed!",
+        )));
+    }
+
+    output("📖", "Getting started...")?;
+
+    let base_is_master = pull_request.base.is_master_branch();
+
+    let result = gh
+        .update_pull_request(
+            pull_request_number,
+            PullRequestUpdate {
+                state: Some(PullRequestState::Closed),
+                ..Default::default()
+            },
+        )
+        .await;
+
+    match result {
+        Ok(closed) => closed,
+        Err(error) => {
+            output("❌", "GitHub Pull Request close failed")?;
+
+            return Err(error);
+        }
+    };
+
+    output("📕", "Closed!")?;
+
```

Comment:
At this point it would be nice to rewrite the local commit's message and remove the "Pull Request:" section. That would mean that you could open a new PR immediately with `spr diff`. Also it's just "more correct", you know...

I haven't tried it, but it should be easy, something along the lines of

```
prepared_commit.message.remove(&MessageSection::PullRequest);
drop(prepared_commit);
git.rewrite_commit_messages(prepared_commits.as_mut_slice(), None)?;
```

Reply:
Sounds good! Lemme try and implement that.

Reply:
So, I'm very much a Rust novice and what you showed worked. But I'm curious, why do we need the `drop`? From what I've read about `drop` and how it works, it seems like since the `prepared_commit` goes out of scope, it shouldn't be a big deal. But I recognize that I'm probably missing some stuff here.

The main reason I ask is that after adding the `--all` option and extracting a function (to make things a little more understandable/similar) the `drop` seemed a little superfluous and I left it out. But can easily add it back in if that's the right way.

Reply:
I can't say with confidence either. Did you try it without the `drop`? Maybe it works.

My thinking was that `prepared_commit` borrows something from within `prepared_commits`, so I wanted to end that before using `prepared_commits`. But maybe Rust figures that out without `drop`?

Not that we need to know now, with the changes above the `drop` isn't needed anymore anyway.

Reply:
Yeah, it did work without the `drop`. I just assumed it was one of those things that type checks fine, but could be a potential issue for some reason I'm not familiar with.

Reply:
Sorry, just had a thought that was related to the original comment:

> At this point it would be nice to rewrite the local commit's message and remove the "Pull Request:" section. That would mean that you could open a new PR immediately with `spr diff`. Also it's just "more correct", you know...

What about the "Reviewers" and "Reviewed By" sections? If the idea is that the "Pull Request" section is no longer part of the commit, then it seems just as reasonable to remove the "Reviewed By" section.

The "Reviewers" section I'm not too sure about. On the on hand, you likely need to change what's in the commit if it's being closed, so you might need a different set of reviewers. Plus, choosing reviewers doesn't work for people without write access to the repo. So it'd cause the next `spr diff` to fail for sure. On the other hand, you can set the reviewers yourself irrespective of the PR.

What's your take on each of these sections?

Reply:
Good points. Yes, "Reviewed By" should be removed.

I would leave the "Reviewers" untouched, so they act as the default choice if you resubmit. You can edit the commit message if you want to change that choice. I don't have strong opinions on this one, if you strongly feel "Reviewers" should go, I wouldn't oppose that either.

With regards to the "Reviewers" for people without write access: we should fix that in spr diff, so that it's merely a warning. If you put in a "Reviewers" section, you expect spr to set those for the PR, so if it can't, it should warn you about it, but it shouldn't appear as a failure.

Reply:
That makes sense.

Also, if you follow my line of reasoning further:

> On the on hand, you likely need to change what's in the commit if it's being closed, so you might need a different set of reviewers

you end up thinking that maybe the "Summary" and "Test Plan" sections should be removed as well. But that'd be a bit much, I think. So yeah, removing the "Reviewers" section is probably unwarranted.